File: dp/pascals_triangle.py
# Problem 118: Pascal's Triangle
# Approach: Simulation / Dynamic Programming
# Time Complexity: O(numRows²)
# Space Complexity: O(numRows²)


class Solution(object):
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        ans=[[1]]
        for i in range(numRows-1):
            t=[0]+ans[-1]+[0]
            r=[]
            for j in range(len(ans[-1])+1):
                r.append(t[j]+t[j+1])
            ans.append(r)
        return ans

        

File: dp/pascals_triangle.cpp
// Problem 118: Pascal's Triangle
// Approach: Simulation / Dynamic Programming
// Time Complexity: O(numRows²)
// Space Complexity: O(numRows²)

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>>ans={{1}};
        for(int i=0;i<numRows-1;i++)
        {
            vector<int>t,r,p=ans.back();
            t.push_back(0);
            t.insert(t.end(),p.begin(),p.end());
            t.push_back(0);
            for(int j=0;j<ans.back().size()+1;j++)
            {
                r.push_back(t[j]+t[j+1]);
            }
            ans.push_back(r);
        }
        return ans;
    }
};
